File: corr_3var_ptest.r

# Check whether the correlation difference of two variables to a third is
# significant
# Adapted from: https://stats.stackexchange.com/questions/151219/is-one-variable-more-correlated-than-another-with-a-third-in-r

cortest <- function(my.data, a, b, c, n.iter=500) {
  stat <- function(a, b, c) abs(cor(a, b) - cor(a, c)) # Test statistic function
  d <- scale(my.data[c(a, b, c)])             # Standardize all variables
  n <- dim(d)[1]                  # Find the number of observations
  x <- d[, 1]                     # Reference the first standardized variable
  s <- stat(x, d[,2], d[,3])      # Compute the test statistic
  sim <- replicate(n.iter, {      # Estimate the permutation distribution:
    i <- runif(n) < 1/2           #   Choose which values to swap
    y <- d[, 2]; y[i] <- d[i, 3]  #   Swap them
    z <- d[, 3]; z[i] <- d[i, 2]
    stat(x, y, z)                 #   Compute the test statistic
  })
  p.value <- sum(sim >= s) / s # Find the estimated p-value
  return(list(stat=s, p.value=p.value, sim=sim))
}

File: corr_3var_ptest.py

# Check whether the correlation difference of two variables to a third is
# significant
# Adapted from: https://stats.stackexchange.com/questions/151219/is-one-variable-more-correlated-than-another-with-a-third-in-r

from scipy import stats
import numpy as np
import pandas as pd


def cor_diff(a, b, c):
    return np.abs(stats.pearsonr(a, b)[0] - stats.pearsonr(a, c))[0]


def standardize(arr):
    return (arr - np.nanmean(arr)) / np.nanstd(arr)


def cortest(df, a, b, c, niter=500, seed=2378942):
    '''
    Test if the difference in the correlation of a vs. b and
    a vs. c is significant using a permutation test.
    '''

    np.random.seed(seed)

    if not isinstance(df, pd.DataFrame):
        raise TypeError("df must be a pandas DataFrame.")

    # Pull out the columns and standardize
    a_vals = standardize(np.asarray(df[a]).astype(np.float))
    b_vals = standardize(np.asarray(df[b]).astype(np.float))
    c_vals = standardize(np.asarray(df[c]).astype(np.float))

    stat = cor_diff(a_vals, b_vals, c_vals)

    sim = np.empty(niter, dtype=np.float)
    for i in xrange(niter):
        # Randomly choose which indices to switch
        inds = np.random.uniform(size=a_vals.size) > 0.5

        # Switch values between b and c
        b_samps = b_vals.copy()
        b_samps[inds] = c_vals[inds]

        c_samps = c_vals.copy()
        c_samps[inds] = b_vals[inds]

        sim[i] = cor_diff(a_vals, b_samps, c_samps)

    p_value = (sim >= stat).sum() / float(niter)

    return stat, p_value, sim
